File: Tests/ChatUnpackCoreTests/TestSupport.swift
import Foundation
import CoreGraphics
import ChatUnpackCore

struct TestSuite {
  private(set) var checkCount = 0
  private(set) var failures: [String] = []

  mutating func expect(
    _ condition: @autoclosure () -> Bool,
    _ name: String,
    file: String = #fileID,
    line: Int = #line
  ) {
    checkCount += 1
    guard !condition() else { return }
    failures.append("\(file):\(line) \(name)")
  }

  func finish() -> Never {
    if failures.isEmpty {
      print("核心测试通过：\(checkCount) 项检查")
      exit(0)
    }

    fputs("核心测试失败：\(failures.count)/\(checkCount)\n", stderr)
    failures.forEach { fputs("- \($0)\n", stderr) }
    exit(1)
  }
}

func makeMessage(
  sender: String = "测试用户",
  timestamp: String = "2026年8月21日 09:51",
  body: [String] = ["测试消息"],
  senderConfidence: Float = 0.99,
  timestampConfidence: Float = 0.99,
  bodyConfidence: Float = 0.99,
  kind: MessageKind = .text,
  isPartial: Bool = false
) -> ChatMessage {
  ChatMessage(
    sender: RecognizedField(text: sender, confidence: senderConfidence),
    timestamp: RecognizedField(text: timestamp, confidence: timestampConfidence),
    body: body.map { RecognizedLine(text: $0, confidence: bodyConfidence) },
    kind: kind,
    isPartial: isPartial
  )
}

func makeOCRLine(
  _ text: String,
  x: CGFloat,
  top: CGFloat,
  width: CGFloat = 0.2,
  height: CGFloat = 0.04,
  confidence: Float = 0.99,
  viewportIndex: Int = 0
) -> OCRLine {
  OCRLine(
    text: text,
    confidence: confidence,
    boundingBox: CGRect(
      x: x,
      y: 1 - top - height,
      width: width,
      height: height
    ),
    viewportIndex: viewportIndex
  )
}

File: windows/tests/ChatUnpack.Core.TestRunner/TestSupport.cs
using System.Runtime.CompilerServices;

using ChatUnpack.Core.Domain;

namespace ChatUnpack.Core.TestRunner;

internal sealed class TestSuite
{
  private readonly List<string> failures = new();

  public int CheckCount { get; private set; }

  public void Expect(
    bool condition,
    string name,
    [CallerFilePath] string file = "",
    [CallerLineNumber] int line = 0)
  {
    CheckCount += 1;
    if (!condition)
    {
      failures.Add($"{Path.GetFileName(file)}:{line} {name}");
    }
  }

  public int Finish()
  {
    if (failures.Count == 0)
    {
      Console.WriteLine($"核心测试通过：{CheckCount} 项检查");
      return 0;
    }

    Console.Error.WriteLine($"核心测试失败：{failures.Count}/{CheckCount}");
    foreach (var failure in failures)
    {
      Console.Error.WriteLine($"- {failure}");
    }

    return 1;
  }
}

internal static class TestData
{
  public static ChatMessage MakeMessage(
    string sender = "测试用户",
    string timestamp = "2026年8月21日 09:51",
    IEnumerable<string>? body = null,
    double? senderConfidence = 0.99,
    double? timestampConfidence = 0.99,
    double? bodyConfidence = 0.99,
    MessageKind kind = MessageKind.Text,
    bool isPartial = false)
  {
    return new ChatMessage(
      new RecognizedField(sender, senderConfidence),
      new RecognizedField(timestamp, timestampConfidence),
      (body ?? ["测试消息"]).Select(text => new RecognizedLine(text, bodyConfidence)),
      kind,
      isPartial: isPartial);
  }

  public static OcrLine MakeOcrLine(
    string text,
    double x,
    double top,
    double width = 0.2,
    double height = 0.04,
    double? confidence = 0.99,
    IEnumerable<string>? alternatives = null,
    int viewportIndex = 0)
  {
    return new OcrLine(
      text,
      confidence,
      new NormalizedRect(x, 1 - top - height, width, height),
      alternatives,
      viewportIndex);
  }
}
